server/src/main/java/io/crate/execution/ddl/tables/TableModifyTask.java

```diff
@@ -0,0 +1,78 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.execution.ddl.tables;
+
+import java.io.IOException;
+
+import org.elasticsearch.action.support.master.AcknowledgedRequest;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.common.TriFunction;
+import org.elasticsearch.index.mapper.MapperService;
+
+import io.crate.common.CheckedFunction;
+import io.crate.metadata.NodeContext;
+import io.crate.metadata.RelationName;
+import io.crate.metadata.cluster.DDLClusterStateTaskExecutor;
+import io.crate.metadata.doc.DocTableInfo;
+import io.crate.metadata.doc.DocTableInfoFactory;
+
+public class TableModifyTask<T extends AcknowledgedRequest<?>> extends DDLClusterStateTaskExecutor<T> {
+
+    private final NodeContext nodeContext;
+    private final CheckedFunction<IndexMetadata, MapperService, IOException> createMapperService;
+    private final RelationName relationName;
+    private final TableModifyExecutor<T> executor;
+
+    public TableModifyTask(NodeContext nodeContext,
+                           CheckedFunction<IndexMetadata, MapperService, IOException> createMapperService,
+                           RelationName relationName,
+                           TableModifyExecutor<T> executor) {
+        this.nodeContext = nodeContext;
+        this.createMapperService = createMapperService;
+        this.relationName = relationName;
+        this.executor = executor;
+    }
+
+    @Override
+    protected ClusterState execute(ClusterState currentState, T t) throws Exception {
+        DocTableInfoFactory docTableInfoFactory = new DocTableInfoFactory(nodeContext);
+        Metadata metadata = currentState.metadata();
+        DocTableInfo currentTable = docTableInfoFactory.create(relationName, metadata);
+        Metadata.Builder metadataBuilder = Metadata.builder(metadata);
+        DocTableInfo newTable = executor.apply(t, currentTable, metadataBuilder);
+        if (newTable == currentTable) {
+            return currentState;
+        }
+        newTable.writeTo(createMapperService, metadata, metadataBuilder);
+        Metadata newMetadata = metadataBuilder.build();
+        // Ensure new table can still be parsed
+        docTableInfoFactory.create(relationName, newMetadata);
+        return ClusterState.builder(currentState)
+            .metadata(newMetadata)
+            .build();
+    }
+
+    public interface TableModifyExecutor<T extends AcknowledgedRequest<?>> extends TriFunction<T, DocTableInfo, Metadata.Builder, DocTableInfo> {
```

Comment:
```suggestion
    public interface TableModifyExecutor<T> extends TriFunction<T, DocTableInfo, Metadata.Builder, DocTableInfo> {
```

The additional constraint is not used

Reply:
Thank you for reviewing, please check the commit for updates.